### PR TITLE
feat(engine): migrate service

### PR DESCRIPTION
**Pull request by fsamin**
1. Description
1. Related issues
1. About tests
1. Mentions

@ovh/cds

### CODE REVIEW

File: engine/api/database/dbmigrate/migrate.go [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+	hostname, err := os.Hostname()
+	if err != nil {
+		return nil, err
+	}
+	hostname = fmt.Sprintf("%s-%d", hostname, time.Now().UnixNano())
+	if err := lockMigrate(DBFunc(), hostname); err != nil {
+		return nil, err
+	}
+
+	defer unlockMigrate(DBFunc(), hostname)
```
Comment (golangci) [inline engine/api/database/dbmigrate/migrate.go]: Error return value of `unlockMigrate` is not checked

File: engine/api/database/dbmigrate/migrate.go [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+	if err := dbmap.CreateTablesIfNotExists(); err != nil {
+		return err
+	}
+
+	tx, err := dbmap.Begin()
+	if err != nil {
+		return err
+	}
+
+	defer tx.Rollback()
```
Comment (golangci) [inline engine/api/database/dbmigrate/migrate.go]: Error return value of `tx.Rollback` is not checked

File: engine/api/database/dbmigrate/migrate.go [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+	// construct a gorp DbMap
+	dbmap := &gorp.DbMap{Db: db, Dialect: gorp.PostgresDialect{}}
+	dbmap.AddTableWithName(MigrationLock{}, "gorp_migrations_lock").SetKeys(false, "ID")
+
+	tx, err := dbmap.Begin()
+	if err != nil {
+		return err
+	}
+
+	defer tx.Rollback()
```
Comment (golangci) [inline engine/api/database/dbmigrate/migrate.go]: Error return value of `tx.Rollback` is not checked

File: engine/migrateservice/service.go [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+		log.Info("DBMigrate> Starting HTTP Server on port %d", s.cfg.HTTP.Port)
+		if err := server.ListenAndServe(); err != nil {
+			log.Error("DBMigrate> Listen and serve failed: %s", err)
+		}
+	}()
+
+	//Gracefully shutdown the http server
+	<-c.Done()
+	log.Info("DBMigrate> Shutdown HTTP Server")
+	server.Shutdown(c)
```
Comment (golangci) [inline engine/migrateservice/service.go]: Error return value of `server.Shutdown` is not checked

File: engine/migrateservice/service.go [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+	"time"
+
+	"github.com/gorilla/mux"
+
+	"github.com/ovh/cds/engine/api"
+	"github.com/ovh/cds/engine/api/database"
+	"github.com/ovh/cds/engine/api/database/dbmigrate"
+	"github.com/ovh/cds/engine/api/services"
+	"github.com/ovh/cds/engine/service"
+
```
Comment (bnjjj) [inline engine/migrateservice/service.go]: delete useless blank line

File: engine/migrateservice/service.go [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+			sdk.MonitoringStatusLine{
+				Component: "SQL",
+				Value:     s.currentStatus.err.Error(),
+				Status:    sdk.MonitoringStatusAlert,
+			},
+		)
+		return response
+	}
+
+	var i int
```
Comment (bnjjj) [inline engine/migrateservice/service.go]: this variable name is not very explicit

File: engine/migrateservice/migrate.go [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+	"github.com/rubenv/sql-migrate"
+
+	"github.com/ovh/cds/engine/api/database"
+	"github.com/ovh/cds/engine/api/database/dbmigrate"
+	"github.com/ovh/cds/sdk/log"
+)
+
+func (s *dbmigservice) doMigrate() error {
+	log.Info("%+v", s.cfg.DB)
```
Comment (sguiheux) [inline engine/migrateservice/migrate.go]: remove

File: engine/migrateservice/migrate.go [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+		s.cfg.DB.Password,
+		s.cfg.DB.Name,
+		s.cfg.DB.Host,
+		s.cfg.DB.Port,
+		s.cfg.DB.SSLMode,
+		s.cfg.DB.ConnectTimeout,
+		s.cfg.DB.Timeout,
+		s.cfg.DB.MaxConn)
+	if err != nil {
+		return fmt.Errorf("cannot connect to database: %v", err)
```
Comment (sguiheux) [inline engine/migrateservice/migrate.go]: wrap

File: engine/migrateservice/migrate.go [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+	}
+
+	if _, err := dbmigrate.Do(dbConn.DB, s.cfg.Directory, migrate.Up, false, -1); err != nil {
+		return err
```
Comment (sguiheux) [inline engine/migrateservice/migrate.go]: wrap

File: engine/migrateservice/migrate.go [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+		s.cfg.DB.Password,
+		s.cfg.DB.Name,
+		s.cfg.DB.Host,
+		s.cfg.DB.Port,
+		s.cfg.DB.SSLMode,
+		s.cfg.DB.ConnectTimeout,
+		s.cfg.DB.Timeout,
+		s.cfg.DB.MaxConn)
+	if err != nil {
+		return nil, fmt.Errorf("cannot connect to database: %v", err)
```
Comment (sguiheux) [inline engine/migrateservice/migrate.go]: wrap

File: engine/migrateservice/service.go [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+		log.Error("DBMigrate> Migration failed %v", err)
+		s.currentStatus.err = err
+	}
+
+	log.Info("DBMigrate> Retrieving Database migration status...")
+	status, err := s.getMigrate()
+	if err != nil {
+		log.Error("DBMigrate> Migration status unavailable %v", err)
+	}
+	s.currentStatus.err = err
```
Comment (sguiheux) [inline engine/migrateservice/service.go]: do not erase if doMigrate failed

File: engine/service/service.go [hunk 12]
```diff
@@ -29,3 +44,7 @@ type Service interface {
 	DoHeartbeat(status func() sdk.MonitoringStatus) (int, error)
 	Status() sdk.MonitoringStatus
 }
+
+type BeforeStart interface {
```
Comment (sguiheux) [inline engine/service/service.go]: doc

File: engine/types.go [hunk 13]
```diff
@@ -44,9 +45,10 @@ type Configuration struct {
...
-	Hooks        hooks.Configuration        `toml:"hooks" comment:"######################\n CDS Hooks Settings \n######################"`
-	VCS          vcs.Configuration          `toml:"vcs" comment:"######################\n CDS VCS Settings \n######################"`
-	Repositories repositories.Configuration `toml:"repositories" comment:"######################\n CDS Repositories Settings \n######################"`
+	Hooks           hooks.Configuration          `toml:"hooks" comment:"######################\n CDS Hooks Settings \n######################"`
+	VCS             vcs.Configuration            `toml:"vcs" comment:"######################\n CDS VCS Settings \n######################"`
+	Repositories    repositories.Configuration   `toml:"repositories" comment:"######################\n CDS Repositories Settings \n######################"`
+	DatabaseMigrate migrateservice.Configuration `toml:"databaseMigrate" comment:"######################\n CDS DB Migrate Service Settings \n######################"`
 }
 
 func AsEnvVariables(o interface{}, prefix string, skipCommented bool) map[string]string {
```
Comment (sguiheux) [inline engine/types.go]: doc

File: engine/migrateservice/service.go [hunk 14]
```diff
@@ -143,22 +144,22 @@ func (s *dbmigservice) Status() sdk.MonitoringStatus {
 		return response
 	}
 
-	var i int
+	var theNumberOfSuccessfulMigations int
```
Comment (bnjjj) [inline engine/migrateservice/service.go]: sad news ! I can't why you mean by `migations`

**Pull request merged**